examples/complex/block_dns_over_https.py

```diff
@@ -0,0 +1,248 @@
+"""
+This module is for blocking DNS over HTTPS requests.
+
+It loads a blocklist of IPs and hostnames that are known to serve DNS over HTTPS requests.
+It also uses headers, query params, and paths to detect DoH (and block it)
+"""
+import json
+import re
+import os
+import urllib.request
+from typing import List
+
+import dns.query
+import dns.rdatatype
+import dns.message
+import dns.resolver
+import dns.rdtypes.IN.A
+import dns.rdtypes.IN.AAAA
+
+from mitmproxy import ctx
+
+# filename we'll save the blocklist to so we don't have to re-generate it every time
+blocklist_filename = 'blocklist.json'
+
+# additional hostnames to block
+additional_doh_names: List[str] = [
+    'dns.google.com'
+]
+
+# additional IPs to block
+additional_doh_ips: List[str] = [
+
+]
+
+
+def get_doh_providers():
+    """
+    Scrape a list of DoH providers from curl's wiki page.
+    :return: a generator of dicts containing information about the DoH providers
+    """
+    https_url_re = re.compile(r'https://'
+                              r'(?P<hostname>[0-9a-zA-Z._~-]+)'
+                              r'(?P<port>:[0-9]+)?'
+                              r'(?P<path>[0-9a-zA-Z._~/-]+)?')
+
+    provider_re = re.compile(r'(\[([^\]]+)\]\(([^)]+))\)|(.*)')
+    # URLs that are not DoH URLs
+    do_not_include = ['my.nextdns.io', 'blog.cloudflare.com']
+    found_table = False
+    with urllib.request.urlopen('https://raw.githubusercontent.com/wiki/curl/curl/DNS-over-HTTPS.md') as fp:
```

Comment:
Depending on an external website that is not under our control is very brittle.
Please add some error handling and maybe a fallback static list of a handful common providers so this example still works in a few months from now after that wiki page has changed significantly.

Reply:
> Depending on an external website that is not under our control is very brittle.
Very good point. I'll just replace all the wiki scraping and DNS lookups and what with a static list.

Reply:
Alright made the changes.